EZP-31593 Move REST API descriptions from old reference to developer doc (#1128)
ezsystems/developer-documentation: docs/api/rest_api_best_practices.md

# REST API reference best practices

This page refers to [REST API reference](https://doc.ezplatform.com/rest-api-reference), where you can find detailed information about
REST API resources and endpoints.

## Specifying SiteAccess

> In order to specify SiteAccess, when talking to the REST API, provide a custom header `X-Siteaccess`.

Reviewer: ```suggestion
In order to specify a SiteAccess when communicating with the REST API, provide a custom `X-Siteaccess` header.
```

> If it is not provided, the default one will be used.

Reviewer: ```suggestion
If it is not provided, the default SiteAccess is be used.
```

Example:

```
GET / HTTP/1.1
Host: api.example.com
Accept: application/vnd.ez.api.Root+json
X-Siteaccess: admin
```

## Media types

The methods on resources provide multiple media types in their responses.

> A media type can be selected in the `Accept` Header.

Reviewer: ```suggestion
A media type can be selected in the `Accept` header.
```

> Each XML media type has a unique name e.g. `application/vnd.ez.api.User+xml`.

Reviewer: ```suggestion
Each XML media type has a unique name, e.g. `application/vnd.ez.api.User+xml`.
```

> The returned XML response conforms with the complex type definition with a name e.g. `vnd.ez.api.User` in the `user.xsd` XML schema definition file (see `User_`).

Reviewer: ```suggestion
The returned XML response conforms with the complex type definition with a name, e.g. `vnd.ez.api.User` in the `user.xsd` XML schema definition file (see `User_`).
```

  xmlns="http://ez.no/API/Values" targetNamespace="http://ez.no/API/Values">
  <xsd:include schemaLocation="CommonDefinitions.xsd" />
  <xsd:complexType name="vnd.ez.api.MobileContent">
    <xsd:complexContent>
      <xsd:extension base="ref">
        <xsd:all>
          <xsd:element name="Title" type="xsd:string" />
          <xsd:element name="Summary" type="xsd:string" />
        </xsd:all>
      </xsd:extension>
    </xsd:complexContent>
  </xsd:complexType>
  <xsd:element name="MobileContent" type="vnd.ez.api.MobileContent"/>
</xsd:schema>
```

So that:

```
GET /content/objects/23 HTTP/1.1
Accept: application/vnd.ez.api.MobileContent+xml
```

Returns:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<MobileContent href="/content/objects/23" media-type="application/vnd.ez.api.MobileContent+xml">
  <Title>Title</Title>
  <Summary>This is a summary</Summary>
</MobileContent>
```

In this specification, only the standard schemas and media types are defined (see `InputOutput_`).
If there is only one media type defined for XML or JSON, it is also possible to specify `application/xml` or `application/json`.

## URIs

> The REST API is designed in such a way that the client doesn't need to  construct any URIs to resources.

Reviewer: ```suggestion
The REST API is designed in such a way that the client doesn't need to construct any URIs to resources.
```

Starting from the root resources (`ListRoot_`) every response includes further links to related resources.
The URIs should be used directly as identifiers on the client side and the client should not construct any URIs by using an ID.

### URIs prefix

In [eZ Platform REST reference](https://doc.ezplatform.com/rest-api-reference), for the sake of readability, there are no prefixes used in the URIs.

> In practise, the `/api/ezp/v2` prefixes are all REST hrefs.

Reviewer: ```suggestion
In practice, the `/api/ezp/v2` prefixes are all REST hrefs.
```
